pyscaffold/api/helpers.py

```diff
@@ -0,0 +1,372 @@
+# -*- coding: utf-8 -*-
+"""
+Useful functions for manipulating the action list and project structure.
+"""
+from __future__ import absolute_import
+
+import functools
+import sys
+from copy import deepcopy
+
+from six import string_types
+
+from ..exceptions import ActionNotFound
+from ..log import logger
+from ..structure import FileOp, define_structure
+
+logger = logger  # Sphinx workaround to force documenting imported members
+"""Logger wrapper, that provides methods like :obj:`~.ReportLogger.report`.
+See :class:`~.ReportLogger`.
+"""
+
+NO_OVERWRITE = FileOp.NO_OVERWRITE
+"""Do not overwrite an existing file during update
+(still created if not exists)
+"""
+
+NO_CREATE = FileOp.NO_CREATE
+"""Do not create the file during an update"""
+
+
+# -------- Project Structure --------
+
+def ensure(structure, path, content=None, update_rule=None):
+    """Ensure a file exists in the representation of the project tree
+    with the provided content.
+    All the parent directories are automatically created.
+
+    Args:
+        structure (dict): project representation as (possibly) nested
+            :obj:`dict`. See :obj:`~.merge`.
+        path (str or list): file path relative to the structure root.
```

Comment:
My experience is that functions that have arguments allowing varying data types end up to be complicated and are hard to test. One could think about forcing the user to provide an actual path as string (or `pathlib.Path` object) and telling him in the doc that he/she can easily use `os.path.join` to transform a list of directories into a proper path string (or with the corresponding `pathlib` function in the latter case).

Reply:
Do you think that it is ok to use the `pathlib` for the internal implementation and add the backport for Python 2 as a requirement for PyScaffold?

Reply:
Yes, I think that is okay. Most people will start projects that are Python3-only anyway these days, I guess.